src/network/networkd-dhcp-static-lease.c

```diff
@@ -0,0 +1,231 @@
+        int r;
+
+        assert(network);
+        assert(ret);
+        assert(!!filename == (section_line > 0));
+
+        if (filename) {
+                r = network_config_section_new(filename, section_line, &n);
+                if (r < 0)
+                        return r;
+
+                static_lease = ordered_hashmap_get(network->dhcp_static_leases_by_section, n);
+                if (static_lease) {
+                        *ret = TAKE_PTR(static_lease);
+                        return 0;
+                }
+        }
+
+        r = dhcp_static_lease_new(&static_lease);
+        if (r < 0)
+                return r;
+
+        static_lease->network = network;
+
+        if (filename) {
+                static_lease->section = TAKE_PTR(n);
+                r = ordered_hashmap_ensure_allocated(&network->dhcp_static_leases_by_section, &network_config_hash_ops);
+                if (r < 0)
+                        return r;
+
+                r = ordered_hashmap_put(network->dhcp_static_leases_by_section, static_lease->section, static_lease);
+                if (r < 0)
+                        return r;
+        }
+
+        *ret = TAKE_PTR(static_lease);
+
+        return 0;
+}
+
+void network_drop_invalid_static_leases(Network *network) {
+        DHCPStaticLease *static_lease;
+
+        assert(network);
+
+        ORDERED_HASHMAP_FOREACH(static_lease, network->dhcp_static_leases_by_section)
+                if (dhcp_static_lease_section_verify(static_lease) < 0)
+                        dhcp_static_lease_free(static_lease);
+}
+
+int config_parse_dhcp_static_lease_address(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue, 
+                void *data,
+                void *userdata) {
+
+        Network *network;
+        _cleanup_(dhcp_static_lease_free_or_set_invalidp) DHCPStaticLease *current_lease = NULL;
+        union in_addr_union addr;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(userdata);
+
+        network = userdata;
+
+        r = lease_new_static(network, filename, section_line, &current_lease);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "Failed to create a new static lease: %s", rvalue);
+                return 0;
+        }
+
+        if (isempty(rvalue)) {
+                sd_dhcp_static_lease_unset_address(current_lease->static_lease);
+                return 0;
+        }
+
+        r = in_addr_from_string(AF_INET, rvalue, &addr);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Failed to parse DHCPv4 IPv4 address data, ignoring static lease assignment: %s", rvalue);
```

Comment:
`"Failed to parse IPv4 address for DHCPv4 static lease, ignoring assignment: %s"`